lib/Core/Executor.cpp

```diff
@@ -905,14 +905,21 @@ Executor::fork(ExecutionState &current, ref<Expr> condition, bool isInternal) {
       }
     }
 
+    // We don't add a constraint explicitly yet
```

Comment:
Hi @MartinNowack , I completely understand now your rationale, but I would say that it is completely acceptable for BFS to never exit the loop here.  In the end, it boils down to when we call back into the scheduler.  I think calling the scheduler (or more precisely informing it of changes) when we fork execution into two states is the right thing to do.  In the presence of infinite loops, it's perfectly acceptable for a heuristic to get stuck in the loop.  In fact, if we modify the example to have an infinite loop based on only concrete data (and note that a valid optimization is also to concretize values if they end up being constrained to a single value), you would end up getting stuck in the loop with this code too.  You could further change KLEE's behaviour to reschedule states every instruction, or every N instructions, but I think it's better to stick with the current simple way of doing this after a 2-way fork.  Do you see my point?

@wuestholz , you might be interested in this discussion too.  Do you see my point as to why BFS could loop forever in that second example?